### PR TITLE
[Core] Correcting and optimizing tetrahedra splitting

### PR DESCRIPTION
**Pull request by rubenzorrilla**
We've been debugging for a month to arrive to this PR. Since the majority of the changes proposed appear to be disconnected between them, Long story short, this is how we've arrived to here:

1. At the very beginning we realized that with high pressures, our discontinuous FE space embedded formulation were crashing for high pressure values. Since in principle the pressure value is not a problem, we start to debug the weak BC imposition (PR  #3703 comes from here).
2. It turned to be that the imposition was right but some spurious fluxes used to appear in the interface, even for the case in where we set a solution inside our discontinuous FE space (basically a constant solution).
3. Thanks to this, we noted that the problem was in the splitting patterns, which were no consistent between faces since it was done considering the local elemental ids. instead of the global ones.
4. Having a look to such splitting patterns, we discovered that the `split_tetrahedra.h`had some splitting patterns which were non optimal since they introduced an internal (also called Steiner node) to perform the splitting even though it was no needed. Since such Steiner never appears when the splitting is associated to a level set (instead to be associated to a remeshing process) we couldn't keep it as it was.

Having this in mind, the summary of the changes is:

- Modification of the __`split_tetrahedra.h`__ splitting patterns. We have regenerated all the possible splitting patterns by giving the external faces to TetGen to obtain the internal subdivisions.
- Now both the __`divide_triangle_2d_3.cpp`__ and the __`divide_tetrahedra_3d_4.cpp`__utilities consider the global ids. to decide the edge collapse before the splitting. This enforces the splitting to be consistent between shared faces.
- New tests have been created and the existent ones that have been influenced by the changes above have been updated.

We strongly believe that the current splitting patterns are OK, but since this is a quite important thing that can silently introduce bugs for all the community (@KratosMultiphysics/all ), we kindly ask you to notify any strange behavior that you observe from now on. We have run (and updated) the FluidDynamicsApplication tests, since this application widely uses these utilities, and the obtained values slightly differ (minor differences can be expected).

### CODE REVIEW

File: kratos/tests/cpp_tests/utilities/test_split_tetrahedra.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    aux_ids[10] = -1; // Internal (Steiner) node not required
+
+    TetrahedraSplit::TetrahedraSplitMode(aux_ids.data(), edge_ids.data());
+
+    KRATOS_CHECK_EQUAL(edge_ids[0], 0);
+    KRATOS_CHECK_EQUAL(edge_ids[1], 5);
+    KRATOS_CHECK_EQUAL(edge_ids[2], 3);
+    KRATOS_CHECK_EQUAL(edge_ids[3], 7);
+    KRATOS_CHECK_EQUAL(edge_ids[4], 3);
+    KRATOS_CHECK_EQUAL(edge_ids[5], 9);
```
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_split_tetrahedra.cpp]: minor comment, I think all those macros increase the compilation time (pls correct me if I am wrong @roigcarlo )

alternative solution:
~~~c
const std::vector<int> expected_results{0,5,3,7,3,9 }; // maybe constexpr, not sure
for (std::size_t i=0; i<6; ++i) {
    KRATOS_CHECK_EQUAL(edge_ids[i], expected_results[i]);
}
~~~

this way you only have to compile the macro once

**Comment by ddiezrod** [comment]
@rubenzorrilla I understand that the problem appeared only when using discontinuous distances right?

File: kratos/tests/cpp_tests/utilities/test_split_tetrahedra.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/tests/cpp_tests/utilities/test_split_tetrahedra.cpp]: Good to know! Thanks @philbucher

**Comment by rubenzorrilla** [comment]
@ddiezrod the problem appears in any case you require conformity between split faces (independently of the nature of the level set). Note that the splitting always consider the elemental distances.

File: applications/FluidDynamicsApplication/tests/EmbeddedCouette2DTest/reference_couette_embedded_slip_2D.csv [hunk 2]
```diff
@@ -1,73 +1,73 @@
 #ID, VELOCITY_X, VELOCITY_Y
 1, 0.0, 0.0
 2, 0.0, 0.0
-3, 0.37746382305736065, 0.0
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/tests/EmbeddedCouette2DTest/reference_couette_embedded_slip_2D.csv:4]: ok but: wasn't this supposed to give analytic result?

**Review by RiccardoRossi — COMMENTED** [review]
I am really looking forward this PR since it corrects a long standing implementation error (my error :-D)
and i would approve it straight away, however ... @ddiezrod this correction should be also beneficial although not essential for the TwoFluidSolver. It may also potentially change some results so you should beware...

File: applications/FluidDynamicsApplication/tests/EmbeddedCouette2DTest/reference_couette_embedded_slip_2D.csv (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/tests/EmbeddedCouette2DTest/reference_couette_embedded_slip_2D.csv:4]: Fair question... We can further investigate it, but I'd say that it is associated to the weak imposition of the BC.

**Comment by RiccardoRossi** [comment]
@ddiezrod for the two-fluid case it was not an issue since only pressure was enriched and it does not need continuity. Even in this conditions the fix might be beneficial for your cases

**Comment by rubenzorrilla** [comment]
@RiccardoRossi  approving then? Or we want to further test it? During these days I've run one of @soudah tests and it seems that there are no splitting related problems.

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**